ERCS/erc-7589.md

```diff
@@ -0,0 +1,480 @@
+some use cases.
+
+```solidity
+/// @title ERC-7589 Semi-Fungible Token Roles, optional single transaction extension
+/// @dev See https://eips.ethereum.org/EIPS/eip-7589
+/// Note: the ERC-165 identifier for this interface is 0x5c3d7d74.
+interface ICommitTokensAndGrantRoleExtension /* is IERC7589 */ {
+    /// @notice Commits tokens and grant role in a single transaction.
+    /// @param _grantor The owner of the SFTs.
+    /// @param _tokenAddress The token address.
+    /// @param _tokenId The token identifier.
+    /// @param _tokenAmount The token amount.
+    /// @param _role The role identifier.
+    /// @param _grantee The recipient the role.
+    /// @param _expirationDate The expiration date of the role.
+    /// @param _revocable Whether the role is revocable or not.
+    /// @param _data Any additional data about the role.
+    /// @return commitmentId_ The identifier of the commitment created.
+    function commitTokensAndGrantRole(
+        address _grantor,
+        address _tokenAddress,
+        uint256 _tokenId,
+        uint256 _tokenAmount,
+        bytes32 _role,
+        address _grantee,
+        uint64 _expirationDate,
+        bool _revocable,
+        bytes calldata _data
+    ) external returns (uint256 commitmentId_);
+}
+```
+
+### Role Balance Extension
+
+The core interface allows for querying a token commitment's balance but not for a specific user's balance. To determine
+the total amount of tokens granted to a user, the implementation needs to sum up all the roles granted to that user
+while filtering out any expired roles.
+
+This function was included in an optional extension because it's not always necessary and will likely make the
+implementation much more complex (increasing smart contract risk).
+
+```solidity
+/// @title ERC-7589 Semi-Fungible Token Roles, optional role balance extension
+/// @dev See https://eips.ethereum.org/EIPS/eip-7589
+/// Note: the ERC-165 identifier for this interface is 0x2f35b73f.
+interface IRoleBalanceOfExtension /* is IERC7589 */ {
+    /// @notice Returns the sum of all tokenAmounts granted to the grantee for the given role.
+    /// @param _role The role identifier.
+    /// @param _tokenAddress The token address.
+    /// @param _tokenId The token identifier.
+    /// @param _grantee The user for which the balance is returned.
+    /// @return balance_ The balance of the grantee for the given role.
+    function roleBalanceOf(
+        bytes32 _role,
+        address _tokenAddress,
+        uint256 _tokenId,
+        address _grantee
+    ) external returns (uint256 balance_);
+}
+```
+
+### Metadata Extension
+
+The Roles Metadata extension extends the traditional JSON-based metadata schema of SFTs. Therefore, DApps supporting
+this feature MUST also implement the metadata extension of [ERC-1155](./eip-1155.md). This extension is **optional** and
+allows developers to provide additional information on roles.
+
+Updated Metadata Schema:
```

Comment:
This is JSON Schema I assume? It might be nice to explicitly mention that somewhere.